OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,757 @@
+// A single submission (VkQueueSubmit) can contain multiple `SubmitInfo
+struct SubmitInfo {
+  std::vector<SubmittedCommandBuffer> command_buffers;
+};
+
+// Wraps up all the data that needs to be persistent upon a submission (VkQueueSubmit).
+// `completed_markers` are all the debug markers, that got completed (via "End") within this
+// submission. Their "Begin" might still be in a different submission.
+struct QueueSubmission {
+  SubmissionMetaInformation meta_information;
+  std::vector<SubmitInfo> submit_infos;
+  std::vector<MarkerState> completed_markers;
+  uint32_t num_begin_markers = 0;
+};
+
+}  // namespace internal
+
+/*
+ * This class ultimately is responsible to track command buffer and debug marker timings.
+ * To do so, it keeps tracks of command-buffer allocations, destruction, begins, ends as well as
+ * submissions.
+ * On `VkBeginCommandBuffer` and `VkEndCommandBuffer` it can (if capturing) insert write timestamp
+ * commands (`VkCmdWriteTimestamp`). The same is done for debug marker begins and ends. All that
+ * data will be gathered together at a queue submission (`VkQueueSubmit`).
+ *
+ * Upon every `VkQueuePresentKHR` it will check if the timestamps of a certain submission are
+ * already available, and if so, it will send the results over to the `VulkanLayerProducer`.
+ *
+ * See also `DispatchTable` (for vulkan dispatch), `TimerQueryPool` (to manage the timestamp slots),
+ * and `DeviceManager` (to retrieve device properties).
+ *
+ * Thread-Safety: This class is internally synchronized (using read/write locks), and can be
+ * safely accessed from different threads.
+ */
+template <class DispatchTable, class DeviceManager, class TimerQueryPool>
+class SubmissionTracker : public VulkanLayerProducer::CaptureStatusListener {
+ public:
+  explicit SubmissionTracker(uint32_t max_local_marker_depth_per_command_buffer,
+                             DispatchTable* dispatch_table, TimerQueryPool* timer_query_pool,
+                             DeviceManager* device_manager)
+      : max_local_marker_depth_per_command_buffer_(max_local_marker_depth_per_command_buffer),
+        dispatch_table_(dispatch_table),
+        timer_query_pool_(timer_query_pool),
+        device_manager_(device_manager) {
+    CHECK(dispatch_table_ != nullptr);
+    CHECK(timer_query_pool_ != nullptr);
+    CHECK(device_manager_ != nullptr);
+  }
+
+  // Sets a producer to be used to enqueue capture events and ask if we are capturing.
+  // We will also register our self as CaptureStatusListener, in order to get nofified on
+  // capture finish (OnCaptureFinished). Will reset the open query slots then.
+  void SetVulkanLayerProducer(VulkanLayerProducer* vulkan_layer_producer) {
+    vulkan_layer_producer_ = vulkan_layer_producer;
+    if (vulkan_layer_producer_ != nullptr) {
+      vulkan_layer_producer_->SetCaptureStatusListener(this);
+    }
+  }
+
+  // Sets the maximal depth of debug markers within one command buffer. If set to 0, command buffers
```

Comment:
Do we want "0" to mean "unlimited"? What if I want no debug labels and only command buffer timings? Depth == 1 would still give me one layer of labels, right? 

It seems it would be fine for users to set this to std::numeric_limits<...>::max() if they don't want a depth limit?

Reply:
The comment "If set to 0, command buffers won't be limited by depth" should be clarified. This is still about label depth, right?

Reply:
I changed to std::numeric_limits<uint32_t>::max() and updated the comment.